srcjs/methods/shiny.ts

```diff
@@ -1,28 +1,51 @@
 import { $ } from "../globals";
+import { debounce } from "lodash";
 
-function isBusy(): boolean {
-  if (!$) {
-    return false;
-  }
-  return $("html")
-    .first()
-    .hasClass("shiny-busy");
+let shinyIsIdle = false;
+const shinyIdleFns = [];
+
+function isIdle(): boolean {
+  return shinyIsIdle;
 }
+function callIdleFns() {
+  // call using setTimeout to "break" serialized execution
+  shinyIdleFns.map(function(fn) {
+    setTimeout(fn, 0);
+  });
+}
+
+// wait 200ms before calling callIdleFns
+const callIdleFnsDebounced = debounce(callIdleFns, 200);
 
-function waitUntilIdle(callback, timeout = 23) {
-  const wait = function() {
-    if (isBusy()) {
-      setTimeout(wait, timeout);
-    } else {
-      callback();
-    }
-  };
+if ($) {
+  $(document).on("shiny:busy", function(event) {
+    shinyIsIdle = false;
+    callIdleFnsDebounced.cancel();
+  });
+  $(document).on("shiny:idle", function(event) {
+    shinyIsIdle = true;
+    // to avoid idle and busy thrashing,
```

Comment:
Can you give a short description of a case where busy/idle thrashing causes problems?